### ISSUE TITLE
SMIP: Variable GPU Post Setup User Interaction - WIP

### ISSUE THREAD

**Issue by avive** [spacemeshos/SMIPS#7]
# Core Interactions

## POST SETUP (GUI)
Post commitment Init via a gui app such as Smapp.

1. User selects desired post commitment size. e.g. 100GB, 200GB, etc...
2. User selects path where the post commitment file(s) should be created. This needs to be on a volume that as at least as much free space as the commitment size. The app should verify that this is the case before allowing the user to proceed to step #3.
3. App selects a processor to use for post setup from a list processors available on his system.

    Technical notes:
    - This list always includes the main system CPU as a fall-back processor. It should always be the last item.
    - The list includes each GPU available to the system which is supported by post-gpu. A system may have 0 or more supported GPUs.
    - User should be made aware that using the CPU can be x5-15 times slower then using his GPU.
    - The spacemesh post-gpu c library should be used to enumerate supported system processors. It implements this capability.
    - We need this step to give user full control of the post generation process.
    - Most systems will have 1 supported GPU and 1 CPU. In this case. Select the supported GPU by default as this is what we recommend to use so it is a good default.

4. User is notified that the process has started and that he can check progress in the same screen. e.g. Smesher screen. The app should also send a desktop notification when post setup init is complete.
    - The post commitment job should be started. Note that is should fallback at times of interactivity and continue at full speed when a user is not interactively using the system.

## POST SETUP (CLI)
Flow when post init component is embedded as a full-node component and not a stand-alone process.

1. Run a full node without smeshing yet.
2. Use an API client such as the Spacemesh CLIWallet to query the full node regarding the system post options.
For example, call a GetAvailableProcessors() to return a numbered list of supported post processors. e.g. 0] Nvidia ModelX, 1] AMD modelY, 2]Intel Authentic CPU.
3. Stop the node and set it with CLI params that specify post commitment size, init file location and an index into the supported processors. This index specified which processor should be used by the node for the post init.

---

#### Using GUI
1. Stop the full node.
2. Delete the post init file.
3. Start the full node w/o smeshing.

#### Using a console-based full node
1. Stop the node process.
2. Delete the post commitment file via the OS.

**Comment by avive** [comment]
@noamnelke @moshababo - as discussed - open for comments now as issue before moving to be a smip.

**Comment by avive** [comment]
@moshababo - we talked about it briefly. Here I assume post-init process is going to be packaged as a stand-alone process which exposes an IPC API. I can't find a product reason to bundle the post init utility in go-spacemesh. This is tight-coupling post init to a full node client and arguably a mix of separate concerns. Having a separate post init utility will also make it easier to test.

**Comment by moshababo** [comment]
There are a few different considerations here:

1. The PoST util cannot be concerned solely with the PoST init, but also with generating and validating proofs. Otherwise, this layer will need to be implemented in the full node client, which doesn’t make sense IMO because it will result in not having a single self-contained util which we can sufficiently maintain/test/ship (like the current Go/CPU-based PoST util). As discussed, this layer need to be re-written due to the new tree-free construction, and will be implemented in Rust (thus gaining an easy integration with `post-gpu`, among other benefits).

2. App communication with the PoST util: bypassing the full node client doesn’t make sense to me. The App isn’t a prerequisite for using the full node client, so we’ll need to implement the communication in the full node client anyway, and make it support 2 different use-cases. Since PoST init and mining in general are part of the full node client responsibilities, I think the App should communicate only with the full node client.

3. It might make sense to communicate with the PoST util via IPC instead of in-process shared object linkage, but mainly for being able to utilize a remote machine resources. The testing setup will actually become more difficult, because we'll need to build/spawn/connect to a separate process (while testing the PoST util in isolation is easy regardless of its API technicalities). Also, using IPC for proofs validations could be problematic, so we might potentially want to go hybrid. I'd say that not having to use `cgo` is an advantage, but we'll need it for `svm` anyway, so it doesn't count.  As discussed, I think that the PoST util should support both modes anyway (like the current util). 
 
---

Other remarks:

4. Current PoST util supports splitting the init data into multiple files. I think we should keep supporting this feature. Having hundreds-of-GB-sized file could introduce difficulties to the user.

5. Current PoST util supports init recovery. I think we should keep supporting this feature. 

6. Current PoST util supports initializations for multiple smesher ids. I think we should keep supporting this feature (If so, deleting the PoST init files requires to maintain the respective smesher id).

7. Deletion of the PoST file/s should not be done via OS, but via the PoST util API.

8. AFAIK the PoST isn’t concerned with the net id, but only with the smesher id.

**Comment by noamnelke** [comment]
Product comment: making users select where the PoST work will be done (GPU/CPU) is a bad idea, IMO. What you describe as the default is the correct choice for 99.99% of users. I suspect that the 0.01% who would need control over this won't use the GUI anyway, but you're now forcing everyone to make a choice that some people will get stuck on because they don't understand and others to just need to press "next" one more time.

Agree with everything @moshababo said, except for the last point - we need to add the net id to the hashed message (what's currently only the miner id) to block miners from re-using init files across different networks.

**Comment by avive** [comment]
@moshababo 

> 1. The PoST util cannot be concerned solely with the PoST init, but also with generating and validating proofs. Otherwise, this layer will need to be implemented in the full node client, which doesn’t make sense IMO because it will result in not having a single self-contained util which we can sufficiently maintain/test/ship (like the current Go/CPU-based PoST util). As discussed, this layer need to be re-written due to the new tree-free construction, and will be implemented in Rust (thus gaining an easy integration with `post-gpu`, among other benefits).

Fait argument although there are some pos to have PoST init utility separate and have proof creation and verification embedded in the node.  So it looks like we are going with PoST util embedded in go-spacemesh.

> 2. App communication with the PoST util: bypassing the full node client doesn’t make sense to me. The App isn’t a prerequisite for using the full node client, so we’ll need to implement the communication in the full node client anyway, and make it support 2 different use-cases. Since PoST init and mining in general are part of the full node client responsibilities, I think the App should communicate only with the full node client.

Yes - via the node's new proposed Smesher API. Please see what's been drafted here: https://github.com/spacemeshos/api/tree/smeshing-methods/proto/spacemesh

> 3. It might make sense to communicate with the PoST util via IPC instead of in-process shared object linkage, but mainly for being able to utilize a remote machine resources. The testing setup will actually become more difficult, because we'll need to build/spawn/connect to a separate process (while testing the PoST util in isolation is easy regardless of its API technicalities). Also, using IPC for proofs validations could be problematic, so we might potentially want to go hybrid. I'd say that not having to use `cgo` is an advantage, but we'll need it for `svm` anyway, so it doesn't count.  As discussed, I think that the PoST util should support both modes anyway (like the current util).

ok. So we recommend full embedding of PoST Util (Init, generate proof, verify proof) in go-spacemesh.

> 
> Other remarks:
> 
> 1. Current PoST util supports splitting the init data into multiple files. I think we should keep supporting this feature. Having hundreds-of-GB-sized file could introduce difficulties to the user.

Of-course we should support this. User specified only 1 path on a volume where to create it. However, the PoST util needs to be smart enough to delete all generated PoST files in that path when requested to delete a commitment.


> 2. Current PoST util supports init recovery. I think we should keep supporting this feature.

Please explain what init recovery means. We definitely want to enable continuing commitment file creation that was stopped for any reason. Is this what you mean by recovery?

> 3. Current PoST util supports initializations for multiple smesher ids. I think we should keep supporting this feature (If so, deleting the PoST init files requires to maintain the respective smesher id).

What is the use case of 1 commitment supporting multiple smesher id? Is there really a reason not to tie a commitment to a specific smesher id and network id? I see no such use case and unless there's a good user story here that gives users value we should not support it. Regarding Smesher ID - this concern should be separated from from an actual post commitment file and generation step. Right now they are mixed with each other and we have several bugs in this area such as a new smesher id automatically being created by the node if there's a problem with the init file. The Smesher id is a node concept and should be set and maintained separately from the existence or the state of a PoST commitment file(s) for that smesher id. So, for example, a smesher should be able to create a new PoST commitment if one got corrupted without having to create a new Smesher id. Also relevant to the stop and start smeshing API capabilities we want to have... 

> 4. Deletion of the PoST file/s should not be done via OS, but via the PoST util API.
Of course - see proposed API linked above to deal with commitment. Nothing should be done via the OS - only via the API but the PoST component need to know to deal with unavailable commitments - due to os changes. e.g. user forgets to turn on an external USB drive with a post commitment when trying to smesh...

> 5. AFAIK the PoST isn’t concerned with the net id, but only with the smesher id.
As Noam commented, we definitely want to tie a net id with a PoST together with the Smesher ID so they can't be reused across different networks.

**Comment by avive** [comment]
> Product comment: making users select where the PoST work will be done (GPU/CPU) is a bad idea, IMO. What you describe as the default is the correct choice for 99.99% of users. I suspect that the 0.01% who would need control over this won't use the GUI anyway, but you're now forcing everyone to make a choice that some people will get stuck on because they don't understand and others to just need to press "next" one more time. If you think it's critical to let users control this from the GUI, don't make it a part of the normal flow, put it somewhere under "advanced" or something (I think it's not worth implementing, even there, because nobody will ever use it).

Reducing user friction and unnecessary steps to setup post is critical and we should constantly thinking about. That said, I wish it was possible in this case but I'm afraid it is not. Here's why:

A typical desktop or laptop will have 3 processors that are able to init post, including the main system CPU (slow fallback) - all desktops ship with internal graphics and modern gaming PCs have for example capable (but quite weak) intel GPUs on the motherboard which we support. Macbook pros also have 2 GPUs - one for the internal screen and one for external screens - we support both.

We can't easily figure out, as far as I can tell which GPU enumerated by the system will yield better performance. The dedicated GPU in desktops - eg. an AMD or Nvidia card will yield the best results over the built-in motherboard GPU. In addition, users may want to use the slower motherboard GPU as this will let them have reasonable performance while allowing them to play 60fps using the main GPU - definitely a strong use case.

So, due to the above I believe we should ask the user what processor to use for PoST. If we can find a way to figure out the most performing one (more research needed) then we should consider making it the default in the app for ease of use and not bother the user with other options - but when setting post via a node only setup via a CLI wallet API client or via node startup flags directly - we need to let the user decide which processor to use.  So bottom line, we need to support this selection feature regardless of the app configuration. @noamnelke - makes sense?

**Comment by y0sher** [comment]
> Product comment: making users select where the PoST work will be done (GPU/CPU) is a bad idea, IMO. What you describe as the default is the correct choice for 99.99% of users. I suspect that the 0.01% who would need control over this won't use the GUI anyway, but you're now forcing everyone to make a choice that some people will get stuck on because they don't understand and others to just need to press "next" one more time. If you think it's critical to let users control this from the GUI, don't make it a part of the normal flow, put it somewhere under "advanced" or something (I think it's not worth implementing, even there, because nobody will ever use it).
> 
> Agree with everything @moshababo said, except for the last point - we need to add the net id to the hashed message (what's currently only the miner id) to block miners from re-using init files across different networks.

I agree with this product wise. maybe we can show the minimal `"Next"` screens we can where we actively ask the user to choose something, then show a detailed screen about whats going to happen like a list ex:
```
PoST
- Directory: ~/post/
- Device: GPU
- Size: 250GB
```
and advanced users will be able to change it there.. (maybe add some little (i) icon to explain each of these)

**Comment by noamnelke** [comment]
@avive we can do a 2 second benchmark and figure out which device is fastest.

**Comment by noamnelke** [comment]
I think you underestimate how easy it is to get users to stop a setup process.

"Hmm... what should I do here? Nevermind, I'll go play Fortnight instead"

**Comment by avive** [comment]
> @avive we can do a 2 second benchmark and figure out which device is fastest.

Yes, this is a good idea which will allow us to select the fastest option by default so we don't need to bother the user with it, at least when we use a high-level api but we do want to give smeshers full control over this as well in some cases. e.g. use my motherboard GPU so I can keep playing at 60fps while post is being done reasonably fast...

**Comment by avive** [comment]
> I think you underestimate how easy it is to get users to stop a setup process.
> 
> "Hmm... what should I do here? Nevermind, I'll go play Fortnight instead"

I don't really - I want them to play Fortnite while their computer is creating a commitment - how cool is that?

**Comment by noamnelke** [comment]
please, please put it under "advanced", or as @y0sher suggested, show the user what we chose for them and let them "customize" it.

**Comment by noamnelke** [comment]
> I don't really - I want them to play Fortnite while their computer is creating a commitment - how cool is that?

I think you totally missed my point. Confronted with a hard choice (e.g. one they don't understand) users will opt to do nothing (stop the setup process).

**Comment by avive** [comment]
> please, please put it under "advanced", or as @y0sher suggested, show the user what we chose for them and let them "customize" it.

Yes, if we can benchmark then we can choose the best one as default and let advanced users changes. pPease also see the commitment related methods in the proposed smesher api (wip / pre design review): https://github.com/spacemeshos/api/blob/smeshing-methods/proto/spacemesh/smesher.proto - to have most flexibility in higher level api client design we need to support what's proposed on the basic api level...
